Zend/zend_language_parser.y

```diff
@@ -308,15 +311,29 @@ top_statement_list:
 	|	%empty { $$ = zend_ast_create_list(0, ZEND_AST_STMT_LIST); }
 ;
 
+/* Name usable in a namespace declaration. */
+namespace_declaration_name:
+		identifier								{ $$ = $1; }
+	|	T_NAME_QUALIFIED						{ $$ = $1; }
+;
+
+/* Name usable in "use" declarations (loading separator forbidden). */
 namespace_name:
 		T_STRING								{ $$ = $1; }
-	|	namespace_name T_NS_SEPARATOR T_STRING	{ $$ = zend_ast_append_str($1, $3); }
+	|	T_NAME_QUALIFIED						{ $$ = $1; }
+;
+
+/* Name usable in "use" declarations (leading separator allowed). */
+legacy_namespace_name:
+		namespace_name							{ $$ = $1; }
+	|	T_NAME_FULLY_QUALIFIED					{ $$ = $1; }
 ;
 
 name:
-		namespace_name								{ $$ = $1; $$->attr = ZEND_NAME_NOT_FQ; }
-	|	T_NAMESPACE T_NS_SEPARATOR namespace_name	{ $$ = $3; $$->attr = ZEND_NAME_RELATIVE; }
-	|	T_NS_SEPARATOR namespace_name				{ $$ = $2; $$->attr = ZEND_NAME_FQ; }
+		T_STRING									{ $$ = $1; $$->attr = ZEND_NAME_NOT_FQ; }
+	|	T_NAME_QUALIFIED							{ $$ = $1; $$->attr = ZEND_NAME_NOT_FQ; }
```

Comment:
One minor oversight in the implementation that should be easy to address (It's easy to forget about. I assume the author will address this, and supporting tokens would generally be assumed to be part of the final implementation of any RFC adding a new token type, for anyone wondering. I've made this mistake in drafts of other RFC implementations I've worked on):

1. Constants should be provided to userland for T_NAME_QUALIFIED, T_NAME_FULLY_QUALIFIED, and T_NAME_RELATIVE
2. The implementation of `char *get_token_type_name` should be updated with the new token types, so that calls from PHP to token_name() (and possibly error messages)

At least 2 parts of ext/tokenizer/tokenizer_data.c need to be updated

```c
void tokenizer_register_constants(INIT_FUNC_ARGS) {
	REGISTER_LONG_CONSTANT("T_THROW", T_THROW, CONST_CS | CONST_PERSISTENT);
	REGISTER_LONG_CONSTANT("T_INCLUDE", T_INCLUDE, CONST_CS | CONST_PERSISTENT);
	REGISTER_LONG_CONSTANT("T_INCLUDE_ONCE", T_INCLUDE_ONCE, CONST_CS | CONST_PERSISTENT);
```

```c
char *get_token_type_name(int token_type)
{
	switch (token_type) {

		case T_THROW: return "T_THROW";
		case T_INCLUDE: return "T_INCLUDE";
		case T_INCLUDE_ONCE: return "T_INCLUDE_ONCE";
```

Currently,

```
php > echo token_name(T_NAME_QUALIFIED);

Warning: Uncaught Error: Undefined constant "T_NAME_QUALIFIED" in php shell code:1
Stack trace:
#0 {main}
  thrown in php shell code on line 1
php > echo json_encode(token_get_all('<?php use Foo\Bar as X;'));
[[389,"<?php ",1],[352,"use",1],[392," ",1],[314,"Foo\\Bar",1],[392," ",1],[336,"as",1],[392," ",1],[311,"X",1],";"]
php > echo token_name(314);
UNKNOWN
```

Reply:
Thanks, I had indeed forgotten about this. Should be fixed now!